diff --git a/usr/src/tools/cw/cw.c b/usr/src/tools/cw/cw.c
index 3bc250dc0c..e8bcbc994a 100644
--- a/usr/src/tools/cw/cw.c
+++ b/usr/src/tools/cw/cw.c
@@ -644,12 +644,23 @@ do_gcc(cw_ictx_t *ctx)
 	cw_op_t op = CW_O_LINK;
 	char *model = NULL;
 	int	mflag = 0;
+	const char *sysroot;
 
 	if (ctx->i_flags & CW_F_PROG) {
 		newae(ctx->i_ae, "--version");
 		return;
 	}
 
+	if ((sysroot = getenv("CW_SYSROOT")) != NULL) {
+		char *arg;
+
+		if (asprintf(&arg, "--sysroot=%s", sysroot) < 0) {
+			nomem();
+		}
+		newae(ctx->i_ae, arg);
+		free(arg);
+	}
+
 	newae(ctx->i_ae, "-fident");
 	newae(ctx->i_ae, "-finline");
 	newae(ctx->i_ae, "-fno-inline-functions");
@@ -1511,12 +1522,19 @@ do_cc(cw_ictx_t *ctx)
 {
 	int in_output = 0, seen_o = 0;
 	cw_op_t op = CW_O_LINK;
+	const char *sysroot;
 
 	if (ctx->i_flags & CW_F_PROG) {
 		newae(ctx->i_ae, "-V");
 		return;
 	}
 
+	if ((sysroot = getenv("CW_SYSROOT")) != NULL) {
+		(void) fprintf(stderr, "%s: no CW_SYSROOT for Studio\n",
+		    progname);
+		exit(2);
+	}
+
 	while (--ctx->i_oldargc > 0) {
 		char *arg = *++ctx->i_oldargv;
 
